sympy/algebras/quaternion.py

```diff
@@ -244,26 +271,49 @@ def to_euler(self, seq):
         if not symmetric:
             a, b, c, d = a - c, b + d, c + a, d - b
 
-        # calculate angles
-        half_sum = atan2(b, a)
-        half_diff = atan2(d, c)
-
-        angle_j = 2 * atan2(sqrt(c * c + d * d), sqrt(a * a + b * b))
-        # alternatively, we can use this to avoid the square root:
-        # angle_2 = acos(2*(a*a + b*b)/(a*a + b*b + c*c + d*d) - 1)
-
-        angle_i = half_sum + half_diff
-        angle_k = half_sum - half_diff
+        if avoid_square_root:
+            n2 = self.norm()**2 if symmetric else 2 * self.norm()**2
+            angles[1] = acos((a*a + b*b - c*c - d*d) / n2)
+        else:
+            angles[1] = 2 * atan2(sqrt(c * c + d * d), sqrt(a * a + b * b))
+
+        # Check for singularities in numerical cases
+        angle_test = angles[1].evalf()
+        case = 0
+        pi_f = pi.evalf()
+        if angle_test.is_Number:
```

Comment:
Changed `is_Number` to `is_number` for now.

I don't think there is any numerical problems though, because either:
1.  `atan2` is called with two positive arguments, which only gives a value between `0` and `pi/2` and then is multiplied by 2.
2. `acos` is called with properly normalized values, which can only give results between `0` and `pi`.

In both cases, the result is assured to be real and between `0` and `pi`. 
This is how it is implemented in the numerical conversion in SciPy for example.

Reply:
If you still prefer the use of `case = Piecewise((1, Eq(angles[1], 0)), (2, Eq(angles[1], pi), (0, True))` though, can you guide me true that? I don't really know what you meant here!

Reply:
How obfuscated will be expressions for which angle[1] is 0 or pi? Perhaps you can just test `is_eq(angle[1], S.Zero):` and `is_eq(angle[1], S.Pi)`?

Reply:
By `is_eq` do you mean `equals`?

Reply:
Since something like `if Eq(x, 10)` will not always evaluate you can't use it (except ackwardly) as a condition, so helpers like `is_eq` were added that can be used as `if is_eq(x, 10)`. `from sympy.core.relational import is_eq`.

Reply:
Alright I see it now! Updated accordingly.